internal/os_updater/emt.go

```diff
@@ -6,26 +6,319 @@
-	panic("unimplemented")
+func (t *EMTDownloader) Download() error {
+	config, err := LoadConfig(configFilePath)
+	if err != nil {
+		fmt.Println("Error loading intel_manageability.conf:", err)
+		return err
+	}
+
+	// Perform source verification
+	if !IsTrustedRepository(t.request.Url, config) {
+		errMsg := fmt.Sprintf("URL '%s' is not in the list of trusted repositories.", t.request.Url)
+		fmt.Println(errMsg)
+		return errors.New(errMsg)
+	}
+
+	fmt.Println("Downloading update from", t.request.Url)
+
+	// Check available space on disk
+	isDiskEnough, err := t.checkDiskSpace()
+	if err != nil {
+		fmt.Println("Error checking disk space:", err)
+		return errors.New(err.Error())
+	}
+
+	if !isDiskEnough {
+		errMsg := "Insufficient disk space."
+		fmt.Println(errMsg)
+		return errors.New(errMsg)
+	}
+
+	fmt.Println("Disk space enough. Proceeding to download the artifact.")
+
+	// Download file
+	err = t.downloadFile()
+	if err != nil {
+		fmt.Println("Error downloading the file:", err)
+		return err
+	}
+
+	fmt.Println("Download completed.")
+
+	return nil
 }
 
-// EMTUpdater is the struct for parameters needed to update the OS for EMT.
-type EMTUpdater struct{}
+// readJWTToken reads the JWT token that is used for accessing RS server.
+func (t *EMTDownloader) readJWTToken() (string, error) {
+	file, err := t.fs.Open(JWTTokenPath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
 
-// Update updates the OS for EMT.
-func (u *EMTUpdater) Update() error {
-	panic("unimplemented")
+	token, err := afero.ReadFile(t.fs, JWTTokenPath)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(token)), nil
+}
+
+// checkDiskSpace checks if there is enough disk space to download the artifacts.
+func (t *EMTDownloader) checkDiskSpace() (bool, error) {
+	// Get available disk space
+	var stat unix.Statfs_t
+	err := t.statfs("/var/cache/manageability/", &stat)
+	if err != nil {
+		fmt.Printf("Error getting disk space: %v\n", err)
+		return false, err
+	}
+	availableSpace := stat.Bavail * uint64(stat.Bsize)
+
+	//Read JWT token
+	token, err := t.readJWTTokenFunc()
+	if err != nil {
+		fmt.Println("Error reading JWT token:", err)
+		return false, err
+	}
+
+	// Check if the token exists
+	if token == "" {
+		errMsg := "empty JWT token"
+		fmt.Println(errMsg)
+		return false, errors.New(errMsg)
+	}
+
+	if t.requestCreator == nil {
+		fmt.Println("requestCreator is nil")
+	}
+	if t.httpClient == nil {
+		fmt.Println("httpClient is nil")
+	}
+	// Create a new HTTP request
+	req, err := t.requestCreator("HEAD", t.request.Url, nil)
+	if err != nil {
+		fmt.Printf("Error creating request: %v\n", err)
+		return false, err
+	}
+
+	// Add the JWT token to the request header
+	req.Header.Add("Authorization", "Bearer "+token)
+
+	// Perform the request
+	resp, err := t.httpClient.Do(req)
+	if err != nil {
+		fmt.Printf("Error performing request: %v\n", err)
+		return false, err
+	}
+	defer resp.Body.Close()
```

Comment:
Should we be checking the status code returned and sending back an error if it's not 200/Success?

Reply:
Added status code checks.